// tertestrial/dist/setup-wizard.js
// Generated by LiveScript 1.5.0
var ref$, bold, cyan, green, fs, inquirer, path, map, sort, requireYaml, cp;
ref$ = require('chalk'), bold = ref$.bold, cyan = ref$.cyan, green = ref$.green;
fs = require('fs');
inquirer = require('inquirer');
path = require('path');
ref$ = require('prelude-ls'), map = ref$.map, sort = ref$.sort;
requireYaml = require('require-yaml');
cp = require('shelljs').cp;
function builtInActionSets(){
  var files, i$, len$, file, content, results$ = [];
  files = sort(
  map(function(it){
    return path.basename(it, path.extname(it));
  })(
  fs.readdirSync(path.join(__dirname, '..', 'actions'))));
  for (i$ = 0, len$ = files.length; i$ < len$; ++i$) {
    file = files[i$];
    content = require(path.join(__dirname, '..', 'actions', file + ".yml"));
    results$.push({
      name: content.name,
      value: file
    });
  }
  return results$;
}
function createBuiltinConfig(fileName){
  fs.writeFileSync('tertestrial.yml', "actions: " + fileName);
  return console.log("\nCreated configuration file " + cyan('tertestrial.yml') + ".\nYou are done with the setup. Happy testing!\n");
}
function createCustomConfiguration(template){
  cp(path.join(__dirname, '..', 'actions', template + ".yml"), 'tertestrial.yml');
  return console.log("\nI have created configuration file " + cyan('tertestrial.yml') + " as a starter.\nPlease adapt it to your project.\n");
}
module.exports = function(){
  var questions;
  console.log(bold('Tertestrial setup wizard\n'));
  console.log('We are going to create a Tertestrial configuration file together.\n');
  questions = {
    message: 'Do you want to use a built-in configuration?',
    type: 'list',
    name: 'built-in',
    choices: [
      {
        name: 'No, I want to build my own custom configuration',
        value: 'no'
      }, new inquirer.Separator()
    ].concat(builtInActionSets())
  };
  return inquirer.prompt(questions).then(function(answers){
    var questions;
    if (answers['built-in'] !== 'no') {
      createBuiltinConfig(answers['built-in']);
      process.exit();
    }
    console.log('\nOkay, creating a custom configuration for you.\n');
    questions = {
      message: 'Which configuration to you want to use as a starting point?',
      type: 'list',
      name: 'built-in',
      choices: builtInActionSets()
    };
    return inquirer.prompt(questions).then(function(answers){
      return createCustomConfiguration(answers['built-in']);
    });
  });
};
