// tertestrial/dist/helpers/is-duplicate-checker.js
// Generated by LiveScript 1.5.0
var execSync;
execSync = require('child_process').execSync;
function getProcessCwd(processId){
  return execSync("lsof -p " + processId + " -d cwd -a -Fn", {
    encoding: 'utf8'
  }).split('\n')[1].slice(1);
}
function getTertestrialProcessIds(){
  var this$ = this;
  return execSync('ps -o command,pid', {
    encoding: 'utf8'
  }).split('\n').slice(1).map(function(it){
    return it.split(/\s+(?=\S*$)/);
  }).map(function(arg$){
    var command, pid;
    command = arg$[0], pid = arg$[1];
    return {
      command: command,
      pid: pid
    };
  }).filter(function(it){
    return it.command === 'tertestrial';
  }).map(function(it){
    return it.pid;
  });
}
module.exports = (function(){
  function isDuplicate(){
    var myPid, myCwd, this$ = this;
    myPid = process.pid.toString();
    myCwd = process.cwd();
    return getTertestrialProcessIds().filter((function(it){
      return it !== myPid;
    })).map(getProcessCwd).filter((function(it){
      return it === myCwd;
    })).length > 0;
  }
  return isDuplicate;
}());
