// text-runner/dist/runners/helpers/output-collector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = __importDefault(require("util"));
/** simulates console.log to collect output from a running action */
class OutputCollector {
    constructor() {
        /** collects the received output */
        this.content = [];
    }
    /** appends to the output with a newline */
    // @ts-ignore
    log(message, ...optionalParams) {
        const stringified = [];
        for (const arg of arguments) {
            if (typeof arg === "string") {
                stringified.push(arg);
            }
            else {
                stringified.push(util_1.default.inspect(arg));
            }
        }
        this.content.push(stringified.join(" ") + "\n");
    }
    /** returns the "log" function to be used by actions */
    logFn() {
        return this.log.bind(this);
    }
    /** returns the currently accumulated output */
    toString() {
        return this.content.join("");
    }
}
exports.OutputCollector = OutputCollector;
