// text-runner/dist/formatters/progress-formatter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cli_progress_1 = __importDefault(require("cli-progress"));
const colorette_1 = __importDefault(require("colorette"));
const path_1 = __importDefault(require("path"));
const print_code_frame_1 = require("../helpers/print-code-frame");
class ProgressFormatter {
    // @ts-ignore: ignore unused variable
    constructor(stepCount, configuration) {
        this.configuration = configuration;
        this.progressBar = new cli_progress_1.default.Bar({
            clearOnComplete: true,
            format: colorette_1.default.green(" {bar}") + " {percentage}% | ETA: {eta}s | {value}/{total}",
            hideCursor: undefined,
            stopOnComplete: true
        }, cli_progress_1.default.Presets.shades_classic);
        this.progressBar.start(stepCount, 0);
    }
    // @ts-ignore: unused parameters
    failed(activity, stepName, err, output) {
        this.progressBar.stop();
        console.log();
        console.log();
        console.log(colorette_1.default.dim(output));
        console.log(colorette_1.default.red(`${activity.file.platformified()}:${activity.line} -- ${err.message}\n`));
        console.log();
        print_code_frame_1.printCodeFrame(console.log, path_1.default.join(this.configuration.sourceDir, activity.file.platformified()), activity.line);
    }
    // @ts-ignore: okay to not use parameters here
    skipped(activity, stepName, output) {
        this.progressBar.increment(1);
    }
    // @ts-ignore: okay to not use parameters here
    success(activity, stepName, output) {
        this.progressBar.increment(1);
    }
}
exports.ProgressFormatter = ProgressFormatter;
