// text-runner/dist/activity-list/extract-activities.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const slugify_1 = __importDefault(require("@sindresorhus/slugify"));
/** returns all activities found in the given AstNodeLists */
function extractActivities(ASTs, activeAttributeName) {
    const result = [];
    for (const AST of ASTs) {
        result.push(...extractFromAST(AST, activeAttributeName));
    }
    return result;
}
exports.extractActivities = extractActivities;
/** returns the activities contained in the given AstNodeList */
function extractFromAST(AST, attrName) {
    const result = [];
    for (const node of AST) {
        if (isActiveBlockTag(node, attrName)) {
            result.push({
                actionName: slugify_1.default(node.attributes[attrName]),
                file: node.file,
                line: node.line,
                nodes: AST.getNodesFor(node)
            });
        }
    }
    return result;
}
/** returns whether the given AstNode marks an active region in the document */
function isActiveBlockTag(node, activeAttributeName) {
    return !!node.attributes[activeAttributeName];
}
