// text-runner/dist/configuration/determine-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const merge_configurations_1 = require("./merge-configurations");
const publications_1 = require("./publications/publications");
const defaultValues = {
    actions: {},
    classPrefix: "textrun",
    defaultFile: "",
    exclude: [],
    fileGlob: "**/*.md",
    formatterName: "detailed",
    keepTmp: false,
    offline: false,
    publications: new publications_1.Publications(),
    sourceDir: process.cwd(),
    useSystemTempDirectory: false,
    workspace: "" // will be populated later
};
/**
 * Loads the configuration from disk and returns it.
 *
 * @param configFilePath path of the config file
 * @param cmdlineArgs arguments received on the command line
 */
function determineConfiguration(configFileData, cmdlineArgs) {
    // merge the configs
    const result = merge_configurations_1.mergeConfigurations(cmdlineArgs, configFileData, defaultValues);
    result["publications"] = publications_1.Publications.fromJSON(result["publications"]).sorted();
    return result;
}
exports.determineConfiguration = determineConfiguration;
