// text-runner/dist/configuration/config-file/load-config-file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const yamljs_1 = __importDefault(require("yamljs"));
/** provides the content of the config file in the standardized format */
function loadConfigFile(filename) {
    if (!filename) {
        return {};
    }
    const fileData = yamljs_1.default.load(filename);
    const result = {
        actions: fileData.actions,
        classPrefix: fileData.classPrefix,
        defaultFile: fileData.defaultFile,
        exclude: fileData.exclude,
        fileGlob: fileData.files,
        formatterName: fileData.format,
        keepTmp: fileData.keepTmp,
        offline: fileData.offline,
        publications: fileData.publications,
        useSystemTempDirectory: fileData.useSystemTempDirectory,
        workspace: fileData.workspace
    };
    return result;
}
exports.loadConfigFile = loadConfigFile;
